test/contrib/test_socks.py

```diff
@@ -87,6 +91,26 @@ def _address_from_socket(sock):
         raise RuntimeError("Unexpected addr type: %r" % addr_type)
 
 
+def _set_up_fake_getaddrinfo(monkeypatch):
+    # Work around https://github.com/urllib3/urllib3/pull/2034
+    # Nothing prevents localhost to point to two different IPs. For example, in the
+    # Ubuntu set up by GitHub Actions, localhost points both to 127.0.0.1 and ::1.
+    #
+    # In case of failure, PySocks will try the same request on both IPs, but our
+    # handle_socks[45]_negotiation functions don't handle retries, which leads either to
+    # a deadlock or a timeout in case of a failure on the first address.
+    #
+    # However, some tests need to exercise failure. We don't want retries there, but
+    # can't affect PySocks retries via its API. Instead, we monkeypatch PySocks so that
+    # it only sees a single address, which effectively disables retries.
+    def fake_getaddrinfo(addr, port, family, socket_type):
```

Comment:
I think the correct behavior is to "disable" PySocks retries altogether by resolving the proxy hostname before PySocks and giving it the already-resolved IP address.

Reply:
So you're saying instead of mocking we instead hand PySocks `host="127.0.0.1"`? I'm fine with that too.

Reply:
Agreed, but this would be a larger change that should not block this pull request, in my opinion.